src/search.coffee

```diff
@@ -6,234 +6,257 @@ url = require 'url'
 
-  counts: (query, callback) ->
-    getCounts = (options, callback) =>
-      options = _.extend {}, options, {query, sources: Search.SOURCES}
-      @search 'Composite', options, (err, result) =>
-        return callback err if err
-        callback null, @extractCounts result
-
-    # Two requests are needed. The first request is to get an accurate
-    # web results count and the second request is to get an accurate count
-    # for the rest of the verticals.
-    #
-    # The 1,000 value comes from empirical data. It seems that after 600
-    # results, the accuracy gets quite consistent and accurate. I picked 1,000
-    # just to be in the clear. It also doesn't matter if there are fewer than
-    # 1,000 results.
-    async.map [{skip: 1000}, {}], getCounts, (err, results) ->
-      return callback err if err
-      callback null, _.extend results[1], _.pick(results[0], 'web')
-
-  extractCounts: (result) ->
-    keyRe = /(\w+)Total$/
-
-    _.chain(result?.d?.results or [])
-      .first()
-      .pairs()
-      .filter ([key, value]) ->
-        # Eg. WebTotal, ImageTotal, ...
-        keyRe.test key
-      .map ([key, value]) ->
-        # Eg. WebTotal => web
-        key = keyRe.exec(key)[1].toLowerCase()
-        value = Number value
-
-        switch key
-          when 'spellingsuggestions' then ['spelling', value]
-          else [key, value]
-      .object()
-      .value()
-
-  verticalSearch: (vertical, verticalResultParser, query, options, callback) ->
-    [callback, options] = [options, {}] if _.compact(arguments).length is 4
-
-    @parallelSearch vertical, _.extend({}, options, {query}), (err, result) ->
-      return callback err if err
-      callback null, verticalResultParser result
-
-  mapResults: (results, fn) ->
-    _.chain(results)
-      .pluck('d')
-      .pluck('results')
-      .flatten()
-      .map fn
-      .value()
+  _parallelSearch: (options, callback) ->
+    allRequestOptions = []
+
+    top = Search.MAX_RESULTS
+    if options.top?
+      top = Number options.top
+      options = _.omit options, 'top'
+
+    allRequestOptions.push _.defaults {
+      count: Math.min Search.PAGE_SIZE, top - offset
+      offset
+    }, options for offset in [(options.offset || 0)...top] by Search.PAGE_SIZE
+
+    async.mapLimit allRequestOptions, @parallelLimit, _.bind(@_executeSearch, this),
+      (err, responses) ->
+        callback err if err?
+
+        data = {
+          count: 0
+          results: []
+        }
+
+        # Avoid duplicates by checking result IDs.
+        existingIds = []
+        responses.forEach (response) ->
+          data.count = response.count if response.count?
+          response.results.forEach (result) ->
+            unless result.id in existingIds
+              data.results.push result
+              existingIds.push result.id
+
+        callback null, data
+
+  counts: (query, options, callback) ->
+    [callback, options] = [options, {}] if _.compact(arguments).length is 2
```

Comment:
Here's a goofy trick. See if you can make sense of it. If it works and you like it, feel free. But you'll see this in newer code.

```coffee
counts: (query..., options, callback) ->
  query = query[0] # or 'default'
  ...
```